blog/tests/views_test.py

```diff
@@ -0,0 +1,115 @@
+"""Tests for views' Blog app."""
+
+from datetime import datetime
+
+from django.contrib.auth.models import User
+from django.test import TestCase
+from django.urls import reverse
+
+from ..models import Post
+
+
+class PostListTest(TestCase):
+    """Tests for Post model."""
+
+    def setUp(self):
+        """Set up non-modifiable objects used by all test methods."""
+        bob = User.objects.create(username='bob')
+        self.post_first = Post.objects.create(
+            author=bob,
+            title='Title Bob',
+            text='Text bob',
+            published_date=datetime(day=8, month=11, year=2018),
+        )
+
+        john = User.objects.create(username='john')
+
+        self.post_second = Post.objects.create(
+            author=john, title='Title John', text='Text John'
+        )
+        self.post_third = Post.objects.create(
+            author=john, title='Title John 2', text='Text John 2'
+        )
+        self.post_third.publish()
+
+    def tearDown(self):
+        """Clean-up test data."""
+        del self.post_first
+        del self.post_second
+        del self.post_third
+
+    def test_post_list(self):
+        """Test show post list on the page."""
+        all_post = list(str(i) for i in Post.objects.all())
+        self.assertTrue(
+            all(
+                i in ['Title Bob', 'Title John', 'Title John 2']
+                for i in all_post
+            )
+        )
+
+        response = self.client.get(reverse('post_list'))
+        self.assertEqual(response.status_code, 200)
+
+        self.assertTemplateUsed(
+            response, 'blog/post_list.html', 'blog/base.html'
+        )
+        self.assertEqual(len(response.context['posts']), 2)
+
+    def test_post_detail(self):
+        """Test post detail."""
+        response = self.client.get(reverse('post_detail', kwargs={'pk': 1}))
+        self.assertEqual(response.status_code, 200)
+
+        self.assertTemplateUsed(
+            response, 'blog/post_detail.html', 'blog/base.html'
+        )
+
+        self.assertIn(b'Title Bob', response.content)
+        self.assertIn(b'Text bob', response.content)
+
+        self.assertEqual(str(response.context['post']), 'Title Bob')
+
+    def test_post_new_and_edit(self):
+        """Test create and edit post."""
+        response = self.client.get(reverse('post_new'))
+        self.assertEqual(response.status_code, 200)
+
+        paul = User.objects.create_superuser(
+            username='paul', password='test', email='e@e.com'
+        )
+        self.client.login(username=paul, password='test')
```

Comment:
you should probably refer `paul.username` and pass:
```suggestion
        self.client.login(username=paul.username, password=paul.password)
```

Reply:
If 
```
        self.client.login(username=paul.username, password='test')
```
test work, but if I change 'test' to paul.password - not work.
```ValueError: Cannot assign "<SimpleLazyObject: <django.contrib.auth.models.AnonymousUser object at 0x7ff4fb47c208>>": "Post.author" must be a "User" instance.```
Look pdb:
```
ipdb> paul.password                                                             
'pbkdf2_sha256$100000$LR6Zc7l2LK7B$oUoaAGmLJsnB9dUqhDbDdOa3o2XPdVPSGoIyzpmkU6I='
```
Not understand how to fix it :(

Reply:
Oh... right, the password is hashed in database. So assign in to some variable and reuse in two places..